Various fixes in the Inventory; also make it a proper tree.

srgvg opened the pull request:
As discussed in https://groups.google.com/forum/#!searchin/ansible-devel/ascii/ansible-devel/mEovLhqWTV0/Oed0z9SdOjQJ

This patch set addresses several small bugs in the inventory. I left each of them as a separate commit, as each of them has a specific issue it addresses by its own, which should be easier to review.

Short summary:
- avoiding several duplicates in both hosts and group lists within the Group object, especially coming from the dir parser.
- finally fixing group depth, by checking depth on all child groups
- fixing the difference between direct parents and _all_ ancestors
- only adding groups to the default _all_ group, that are direct childs of it ( == groups with depth 1)
- don't add child groups to the parent ancestors

I have been running this patch set in my own production for the last couple of days, and didn't encounter any issues. Obviously there might be corner cases I couldn't cover.

The patch to the script parser - as I don't use the script api myself atm -  was tested by dumping my dir+ini inventory in the right format, and loading it through a "cat json" script. I managed to get exact identical results here.There's more on this alternative way on parsing host/group vars, but I'll come back to that later.

The last patch - Inventory: raise error when adding a group that already exists - is more of a failsafe to the Inventory object. I never had any real issue here, but this might trigger possible issues with later, bad patches.

jctanner commented:
@sergevanginderachter unfortunately, this seems to break the inventory directory test ...

```
======================================================================
FAIL: test_dir_inventory (TestInventory.TestInventory)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/jtanner/ansible/test/units/TestInventory.py", line 433, in test_dir_inventory
    assert host_vars == expected_vars
AssertionError: 
>>  assert {'inventory_hostname': 'zeus', 'group_names': ['ungrouped'], 'var_a': '3#4', 'inventory_hostname_short': 'zeus'} == {'inventory_hostname': 'zeus', 'group_names': ['greek', 'major-god', 'ungrouped'], 'var_a': '3#4', 'inventory_hostname_short': 'zeus'}

-------------------- >> begin captured stdout << ---------------------
HOST     VARS={'inventory_hostname': 'zeus', 'group_names': ['ungrouped'], 'var_a': '3#4', 'inventory_hostname_short': 'zeus'}
EXPECTED VARS={'inventory_hostname': 'zeus', 'group_names': ['greek', 'major-god', 'ungrouped'], 'var_a': '3#4', 'inventory_hostname_short': 'zeus'}

--------------------- >> end captured stdout << ----------------------

----------------------------------------------------------------------
Ran 145 tests in 4.765s

FAILED (failures=1)
make: *** [tests] Error 1
```

mpdehaan commented:
So it's a little curious that zeus appeared in ungrouped before, that may have been a bug in the test.

I would suspect him to be a major-god, and greek, but not a member of ungrouped.

So it's ok to change the test, but the presense in "ungrouped" still seems to imply a small buglet that needs to be fixed.

@sergevanginderachter can you take a look at this?  Thanks!

srgvg commented:
@mpdehaan working on it atm; agreed ungrouped should not be there;
I think tests should also test for ancestors, not only direct parents, but will look into that after this issue is solved.

Problem is somewhere how InventoryDir add hosts to a group, which influences Host.groups
Other Group parameters seem to be ok, this is about getting the duplicated infoirmation on group member ship straight.

I didn't encounter this particular behaviour as in my inventory, i don't have hosts defined in multiple ini files. Somehow my patch doesn't take this into account and Host.groups doesn't get updated with all hosts.

This is a brain teaser :)

bcoca commented:
i followed the 'ungrouped' bug on ini hosts file a while back. If the host
first appears in 'default' or w/o a heading in ini config file, it went
into to 'ungrouped', if the host happens to appear in a group later on, it
was never removed from ungrouped.

srgvg commented:
@bcoca yeah, that's what also happens in the InventoryDir

mpdehaan commented:
Some "prune from ungrouped" function at top level might be acceptable.

srgvg commented:
Without any patch, exporting the unit test inventory_dir inventory in a tree-ish structure gives this:

```
$ ansible-inventory --tree -i test/units/inventory_test_data/inventory_dir/
[0] all 
|--[1] zeus 
|--[1] morpheus 
|--[1] thor 
|--[1] zeus 
|--[1] morpheus 
|--[1] zeus 
|--[1] morpheus 
|--[1] thor 
|--[1] zeus 
|--[1] thor 
|--[1] morpheus 
|--[1] zeus 
|--[1] thor 
|--[1] morpheus 
|--[1] ungrouped 
|  |--[2] zeus 
|  |--[2] morpheus 
|  |--[2] thor 
|  |--[2] zeus 
|  |--[2] morpheus 
|  |--[2] thor 
|--[1] greek 
|  |--[2] zeus 
|  |--[2] morpheus 
|  |--[2] zeus 
|  |--[2] morpheus 
|--[1] norse 
|  |--[2] thor 
|  |--[2] thor 
|--[1] major-god 
|  |--[2] zeus 
|  |--[2] thor 
|  |--[2] zeus 
|  |--[2] thor 
|  |--[2] zeus 
|  |--[2] thor 
|--[1] minor-god 
|  |--[2] morpheus 
|  |--[2] morpheus 
|  |--[2] morpheus 
```

The same but with the initial patch set (the commits above here) gives something better, but not totally correct either (this needed latest devel b/c some `ValueError: No escaped character`):

```
[0] all 
|--[1] ungrouped 
|  |--[2] zeus 
|  |--[2] morpheus 
|  |--[2] thor 
|--[1] greek 
|  |--[2] zeus 
|  |--[2] morpheus 
|--[1] norse 
|  |--[2] thor 
|--[1] major-god 
|  |--[2] zeus 
|  |--[2] thor 
|--[1] minor-god 
|  |--[2] morpheus 
```

Basically the `ungrouped` group was wrong, but this is just scratching the surface. A lot more was wrong when looking into all object details.

Getting this right was hard, at least for me :-) too many recursive object relations etc.

I'm working on refactoring the InventoryDir parser, and think, at last, I got it working; the only correct inventory is:

```
[0] all 
|--[1] ungrouped 
|--[1] greek 
|  |--[2] zeus 
|  |--[2] morpheus 
|--[1] norse 
|  |--[2] thor 
|--[1] major-god 
|  |--[2] zeus 
|  |--[2] thor 
|--[1] minor-god 
|  |--[2] morpheus 
```

I needed to update the unit test, to keep in mind that groups member of at least one specific group, should never be in `ungrouped`.

I'll first do some more testing at $WORK, and hope I can push the latest patch in coming week.

srgvg commented:
@jctanner @mpdehaan  I added a commit here to resolve the tests.

I ended up doing some major refactor on how InventoryDir() blends everythng, to make sure all objects and their parameters are properly merged. All referenced objects in host/group lists are now objects within the self inventory, whereas before e.g. a parent_list could contain a reference to an object from another parser, or be an incomplete list.

Also:
- cleanup of ungrouped hosts that might be added to another group later on ( and updated the unit test)
- similar cleanup of the all group, to only contain child groups and not grandchildren

Took me some time to get it right. I've got enough recursion debugging karma to last a couple of weeks now.

I extensively tested this with bot my $WORK inventory as with the unit test inventory_dir inventory, then rebased my commits on latest devel and gave that also a last test here.

Think of the trees when testing!

srgvg commented:
Added a patch, after noticing certain groups where mentioned more than once in the host.groups list.
(see https://groups.google.com/d/msg/ansible-devel/mEovLhqWTV0/6wP6ndz7LVoJ)

taylorbarstow commented:
@sergevanginderachter moving our discussion from mailing list

Looks like we are onto exactly the same thing here.  You changes seem solid and solve all of my issues locally.

I agree that the tests on inventory dir are pretty thin right now.  So I added a few more test cases in addition to the one you already saw (I just sent you a pull request, sergevanginderachter/ansible#2).

This still doesn't represent a complete test of inventory dir.  But its an improvement adding more coverage around the issues discussed here.  I held off on testing variables and parent/child relationships because TestInventory.py is getting pretty big, that test logic is already there for other configs, and the whole thing might be due for a refactor.  But that's beyond the scope of this PR.

I hope this is useful to you.  Please let me know if there is anything important you think I missed or anything else I can do to help.

Thank you for your efforts on this Serge, it's a big improvement.

srgvg commented:
Thanks @taylorbarstow I'll check your extra tests and add them in my tree.
For reference, the mentioned list discussion is at https://groups.google.com/forum/#!topic/ansible-devel/F0mAr4KNA3k

srgvg committed:
Merge pull request #2 from taylorbarstow/inventorytree-tests

Adding unit tests for host groups with inventory dir
